File: app/services/planning_service.py
from __future__ import annotations

from datetime import date, datetime, timedelta

from ..repository import fetch_planning_jobs, fetch_planning_technicians


def _normalize_date_input(raw: str | None) -> str:
    value = str(raw or "").strip()
    try:
        return datetime.strptime(value, "%Y-%m-%d").date().isoformat()
    except Exception:
        return date.today().isoformat()


def _add_days(date_string: str, amount: int) -> str:
    parsed = datetime.strptime(date_string, "%Y-%m-%d").date()
    return (parsed + timedelta(days=amount)).isoformat()


def _extract_city(address: str | None) -> str:
    raw = str(address or "").strip()
    if not raw:
        return ""

    parts = [part.strip() for part in raw.split(",") if part.strip()]
    tail = parts[-1] if parts else raw
    if tail.lower() == "belgium" and len(parts) > 1:
        tail = parts[-2]
    if len(tail) >= 5 and tail[:4].isdigit():
        return tail[5:].strip()
    return tail


def _derive_region(address: str | None) -> str:
    raw = str(address or "")
    digits = "".join(ch for ch in raw if ch.isdigit())
    postal_code = None
    for index in range(0, max(len(digits) - 3, 0)):
        chunk = digits[index : index + 4]
        if len(chunk) == 4:
            postal_code = int(chunk)
            break

    if postal_code is None:
        return _extract_city(address) or "Onbekend"

    if 1000 <= postal_code <= 1299:
        return "Brussels"
    if 1300 <= postal_code <= 1499:
        return "Walloon Brabant"
    if 1500 <= postal_code <= 1999:
        return "Vlaams-Brabant"
    if 2000 <= postal_code <= 2999:
        return "Antwerp"
    if 3000 <= postal_code <= 3499:
        return "Vlaams-Brabant"
    if 3500 <= postal_code <= 3999:
        return "Limburg"
    if 4000 <= postal_code <= 4999:
        return "Liege"
    if 5000 <= postal_code <= 5999:
        return "Namur"
    if 6000 <= postal_code <= 7999:
        return "Hainaut"
    if 8000 <= postal_code <= 8999:
        return "West Flanders"
    if 9000 <= postal_code <= 9999:
        return "East Flanders"
    return _extract_city(address) or "Onbekend"


def _build_jobs(rows: list[dict], selected_date: str) -> list[dict]:
    now = datetime.now()
    jobs: list[dict] = []

    for row in rows:
        start = row.get("scheduled_start")
        end = row.get("scheduled_end") or (start + timedelta(hours=1) if start else None)
        status = str(row.get("status") or "").strip() or "unassigned"
        urgent = "dring" in str(row.get("category") or "").lower() or status == "on_the_way"
        overdue = bool(start and start < now and status not in {"done", "completed", "cancelled"})

        jobs.append(
            {
                "id": row.get("job_id"),
                "client": row.get("client_name") or "-",
                "city": _extract_city(row.get("address_raw")),
                "region": _derive_region(row.get("address_raw")),
                "start": start.strftime("%Y-%m-%dT%H:%M:%S") if start else "",
                "end": end.strftime("%Y-%m-%dT%H:%M:%S") if end else "",
                "technician_id": row.get("technician_id") or "unassigned",
                "status": status,
                "urgent": urgent,
                "overdue": overdue,
                "date": selected_date,
            }
        )

    return jobs


def _build_technicians(rows: list[dict], jobs: list[dict]) -> list[dict]:
    usage: dict[str, int] = {}
    for job in jobs:
        technician_id = str(job.get("technician_id"))
        usage[technician_id] = usage.get(technician_id, 0) + 1

    technicians = []
    for row in rows:
        tech_id = row.get("tg_id")
        active_jobs = usage.get(str(tech_id), 0)
        region = next((job["region"] for job in jobs if str(job.get("technician_id")) == str(tech_id)), "General")
        technicians.append(
            {
                "id": tech_id,
                "name": row.get("full_name") or "-",
                "status": "busy" if active_jobs else "available",
                "region": region,
            }
        )

    if any(job.get("technician_id") == "unassigned" for job in jobs):
        technicians.append(
            {
                "id": "unassigned",
                "name": "Unassigned",
                "status": "attention",
                "region": "All",
            }
        )

    return technicians


def _build_week_overview(selected_date: str, technicians: list[dict], jobs: list[dict]) -> dict:
    end_date = _add_days(selected_date, 6)
    totals = []

    for technician in technicians:
        tech_id = str(technician.get("id"))
        totals.append(
            {
                "technician_id": technician.get("id"),
                "name": technician.get("name"),
                "jobs": sum(1 for job in jobs if str(job.get("technician_id")) == tech_id),
            }
        )

    return {
        "message": f"Week view toont voorlopig een lichte overview van {selected_date} tot {end_date}.",
        "totals": totals,
    }


def get_planning_data(date_input: str | None, view_input: str | None):
    selected_date = _normalize_date_input(date_input)
    selected_view = "week" if view_input == "week" else "day"
    end_date = _add_days(selected_date, 6) if selected_view == "week" else selected_date

    technicians_rows = fetch_planning_technicians()
    jobs_rows = fetch_planning_jobs(selected_date, end_date)

    jobs = _build_jobs(jobs_rows, selected_date)
    technicians = _build_technicians(technicians_rows, jobs)

    return {
        "date": selected_date,
        "view": selected_view,
        "technicians": technicians,
        "jobs": jobs,
        "week": _build_week_overview(selected_date, technicians, jobs) if selected_view == "week" else None,
    }

File: node-app/src/services/planningService.js
const { fetchPlanningJobs, fetchPlanningTechnicians } = require("../repository");

function normalizeDateInput(value) {
  const raw = String(value || "").trim();
  if (!/^\d{4}-\d{2}-\d{2}$/.test(raw)) {
    const today = new Date();
    return today.toISOString().slice(0, 10);
  }
  return raw;
}

function addDays(dateString, amount) {
  const date = new Date(`${dateString}T00:00:00`);
  date.setDate(date.getDate() + amount);
  return date.toISOString().slice(0, 10);
}

function extractCity(address) {
  const raw = String(address || "").trim();
  if (!raw) {
    return "";
  }

  const segments = raw
    .split(",")
    .map((segment) => segment.trim())
    .filter(Boolean);
  const tail = segments[segments.length - 1] || raw;
  const noCountry = /belgium$/i.test(tail) ? segments[segments.length - 2] || tail : tail;
  const cityMatch = String(noCountry).match(/\b\d{4}\s+(.+)$/);

  if (cityMatch?.[1]) {
    return cityMatch[1].trim();
  }

  return String(noCountry).replace(/^\d{4}\s+/, "").trim();
}

function deriveRegionFromAddress(address) {
  const raw = String(address || "");
  const match = raw.match(/\b(\d{4})\b/);
  if (!match) {
    return extractCity(address) || "Onbekend";
  }

  const code = Number(match[1]);
  if (code >= 1000 && code <= 1299) return "Brussels";
  if (code >= 1300 && code <= 1499) return "Walloon Brabant";
  if (code >= 1500 && code <= 1999) return "Vlaams-Brabant";
  if (code >= 2000 && code <= 2999) return "Antwerp";
  if (code >= 3000 && code <= 3499) return "Vlaams-Brabant";
  if (code >= 3500 && code <= 3999) return "Limburg";
  if (code >= 4000 && code <= 4999) return "Liege";
  if (code >= 5000 && code <= 5999) return "Namur";
  if (code >= 6000 && code <= 7999) return "Hainaut";
  if (code >= 8000 && code <= 8999) return "West Flanders";
  if (code >= 9000 && code <= 9999) return "East Flanders";
  return extractCity(address) || "Onbekend";
}

function statusClass(status) {
  const raw = String(status || "").trim();
  return raw || "unassigned";
}

function buildTechnicians(rawTechnicians, jobs) {
  const usage = new Map();

  jobs.forEach((job) => {
    if (!job.technician_id) {
      return;
    }
    usage.set(job.technician_id, (usage.get(job.technician_id) || 0) + 1);
  });

  const technicians = rawTechnicians.map((tech) => {
    const activeCount = usage.get(tech.tg_id) || 0;
    const region = jobs.find((job) => Number(job.technician_id) === Number(tech.tg_id))?.region || "General";

    return {
      id: tech.tg_id,
      name: tech.full_name,
      status: activeCount > 0 ? "busy" : "available",
      region,
      role: tech.role,
      active_jobs: activeCount,
    };
  });

  const hasUnassigned = jobs.some((job) => !job.technician_id);
  if (hasUnassigned) {
    technicians.push({
      id: "unassigned",
      name: "Unassigned",
      status: "attention",
      region: "All",
      role: "queue",
      active_jobs: jobs.filter((job) => !job.technician_id).length,
    });
  }

  return technicians;
}

function buildJobs(rawJobs, date) {
  const now = new Date();

  return rawJobs.map((job) => {
    const start = new Date(job.scheduled_start);
    const end = job.scheduled_end ? new Date(job.scheduled_end) : new Date(start.getTime() + 60 * 60 * 1000);
    const city = extractCity(job.address_raw);
    const region = deriveRegionFromAddress(job.address_raw);
    const technicianId = job.technician_id || null;
    const status = statusClass(job.status);
    const urgent = /dring|urgent/i.test(String(job.category || "")) || status === "on_the_way";
    const overdue = start.getTime() < now.getTime() && !["done", "completed", "cancelled"].includes(status);

    return {
      id: job.job_id,
      client: job.client_name || "-",
      city,
      region,
      start: start.toISOString().slice(0, 19),
      end: end.toISOString().slice(0, 19),
      technician_id: technicianId || "unassigned",
      status,
      urgent,
      overdue,
      address: job.address_raw || "",
      title: job.problem_type || job.work_type || job.category || "Job",
      date,
    };
  });
}

function buildWeekOverview(date, technicians, jobs) {
  const endDate = addDays(date, 6);
  return {
    message: `Week view toont voorlopig een lichte overview van ${date} tot ${endDate}.`,
    totals: technicians.map((tech) => ({
      technician_id: tech.id,
      name: tech.name,
      jobs: jobs.filter((job) => String(job.technician_id) === String(tech.id)).length,
    })),
  };
}

async function getPlanningData(dateInput, viewInput) {
  const date = normalizeDateInput(dateInput);
  const view = viewInput === "week" ? "week" : "day";
  const startDate = date;
  const endDate = view === "week" ? addDays(date, 6) : date;

  const [rawTechnicians, rawJobs] = await Promise.all([
    fetchPlanningTechnicians(),
    fetchPlanningJobs(startDate, endDate),
  ]);

  const jobs = buildJobs(rawJobs, date);
  const technicians = buildTechnicians(rawTechnicians, jobs);

  return {
    date,
    view,
    technicians,
    jobs,
    week: view === "week" ? buildWeekOverview(date, technicians, jobs) : null,
  };
}

module.exports = {
  getPlanningData,
};
